### ISSUE TITLE
Recommended way for authorisation/authentication in Postgraphile?

### ISSUE THREAD

**Issue by codef0rmerz** [graphile/postgraphile#1514]
Since [postgraphile does not provide a hook to run JWT verification prior to running each query/mutation](https://github.com/graphile/postgraphile/issues/1454#issuecomment-892458361) at runtime, I tried to use the traditional approach of using JWT middleware with Express as shown below (note that without `credentialsRequired:false` the Apollo Gateway does not kick start and yells for missing Authorization header):

```
// Apollo Gateway
const app = express();

app.use(
  jwt({
    secret: 'secret',
    algorithms: ['HS256'],
    credentialsRequired: false,
  })
);

const server = new ApolloServer({
    debug: true,
    playground: true,
    schema,
    executor,
    context: ({ req }) => {
      const user = req.user || null;
      return { user };
    },
});
 
const serviceList = [
  {
    name: 'postgraphile-apollo1',
    url: `http://localhost:5001/graphql`,
  },
];
const gateway = new ApolloGateway({
  serviceList,
  buildService({ url }) {
    return new RemoteGraphQLDataSource({
      url,
      willSendRequest({ request, context }) {
        request.http.headers.set(
          'user',
          context.user ? JSON.stringify(context.user) : null
        );
      },
    });
  },
});
```

Here, I stamp the JWT payload on `context` and pass it further down the subgraph via HTTP header. However, this just makes every query/mutation public, meaning it does not abort the query/mutation if Authorisation header was missing.

Any way to read the `user` header in the Subgraph to figure out if Authorisation header exists at Gateway level?
Or any better way?

```
// Federated Service http://localhost:5001/graphql
interface MyIncomingMessage extends IncomingMessage {
  user: { user: string };
}

const app = express();

app.use(
  postgraphile('postgres://foo@localhost:5432/bar', 'public', {
    watchPg: true,
    graphiql: true,
    enhanceGraphiql: true,
    appendPlugins: [FederationPlugin],
    pgSettings: (req: MyIncomingMessage) => {
      return {
        'user.username': req.user?.user,
      };
    },
  })
);
```

**Comment by ben-pr-p** [comment]
If you know that you want to abort all queries / mutations if the authorization header is missing, you can do that at the express level with a middleware. 

Before `app.use(postgraphile(...))`, try:
```typescript
app.use('*', (req, res, next) => {
  if (!req.user) {
    return res.status(403).send('Access denied');
  }
  return next();
})
```

Your `pgSettings` function looks good and right, and will let you access do `current_setting('user.username', true)` in your SQL functions for use in row level security policies, etc.

If you are looking to get access to `req.user` in custom resolvers that you may write, you can do that with `additionalGraphQLContextFromRequest`: https://www.graphile.org/postgraphile/usage-library/#additionalgraphqlcontextfromrequest-function

Let me know if I'm not 100% understanding what you're trying to accomplish!

**Comment by codef0rmerz** [comment]
> If you know that you want to abort all queries / mutations if the authorization header is missing, you can do that at the express level with a middleware.
> 
> Before `app.use(postgraphile(...))`, try:
> 
> ```ts
> app.use('*', (req, res, next) => {
>   if (!req.user) {
>     return res.status(403).send('Access denied');
>   }
>   return next();
> })
> ```
> 
Thanks @ben-pr-p but it prevents Apollo Gateway from starting up. 

> Your `pgSettings` function looks good and right, and will let you access do `current_setting('user.username', true)` in your SQL functions for use in row level security policies, etc.
> 
> If you are looking to get access to `req.user` in custom resolvers that you may write, you can do that with `additionalGraphQLContextFromRequest`: https://www.graphile.org/postgraphile/usage-library/#additionalgraphqlcontextfromrequest-function
> 
> Let me know if I'm not 100% understanding what you're trying to accomplish!

Tried that as well but as soon as the Apollo restarts, it fails with `Couldn't load service definitions for "postgraphile-apollo1" at http://localhost:5001/graphql`

```
     additionalGraphQLContextFromRequest: async (req) => {
        const user = JSON.parse(req.headers.user.toString());
        if (user) {
          return {
            user: user.name,
          };
        } else {
          throw new Error('Invalid Token');
        }
      },
```

**Comment by ben-pr-p** [comment]
Ah, I see. I imagine Apollo Gateway needs to run the introspection queries in order to start up, and our `app.use(...)` solution also blocks the introspection queries. 

Do you have other options for what you can pass to Apollo Gateway to get it to start up? For example, could you pass it a `schema.graphql` file, or potentially a Javascript GraphQLSchema object? 

Could you potentially also keep Postgraphile unprotected, but use `willSendRequest` to throw an error if it's unauthenticated?

I believe your issue is that you want Postgraphile to reject unauthenticated requests except for the introspection query. This is impossible, because the introspection query is just like any other query and not distinguished.

**Comment by codef0rmerz** [comment]
@ben-pr-p Yeah, that's why I was hoping to have a some kind of hook that will run just before the query/mutation where I can verify JWT. I could do that with vanilla graphql via Resolvers.

**Comment by ben-pr-p** [comment]
I would see if you could do something with Envelop to reject on all requests except for the introspection one: https://the-guild.dev/blog/introducing-envelop

Or make the inverse of this: https://github.com/helfer/graphql-disable-introspection. This code disables introspection - you want to disable everything *except* introspection.

Or handle it inside of `willSendRequest`

**Comment by benjie** [comment]
Or use makeWrapResolversPlugin to wrap _every_ resolver (or at least the root ones) with this logic.
